File: scpn-mif-rs/crates/mif-lifecycle/src/capacitor_bank.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
//! Series RLC capacitor-bank Crank-Nicolson integrator and the three
//! analytical-response closed forms.

use crate::types::{
    CapacitorBankSpec, CapacitorBankState, ConstructError, FreeResponseError, RlcRegime, StepError,
};

/// Underdamped voltage closed form.
pub fn analytical_voltage_underdamped(spec: &CapacitorBankSpec, t: f64, v0: f64) -> f64 {
    let alpha = spec.damping_factor();
    let omega0 = spec.resonant_frequency();
    let omega_d = (omega0 * omega0 - alpha * alpha).sqrt();
    let damped = (-alpha * t).exp();
    let omega_t = omega_d * t;
    v0 * damped * (omega_t.cos() + (alpha / omega_d) * omega_t.sin())
}

/// Underdamped current closed form.
pub fn analytical_current_underdamped(spec: &CapacitorBankSpec, t: f64, v0: f64) -> f64 {
    let alpha = spec.damping_factor();
    let omega0 = spec.resonant_frequency();
    let omega_d = (omega0 * omega0 - alpha * alpha).sqrt();
    (v0 / (omega_d * spec.inductance_h)) * (-alpha * t).exp() * (omega_d * t).sin()
}

/// Critically damped voltage closed form.
pub fn analytical_voltage_critically_damped(spec: &CapacitorBankSpec, t: f64, v0: f64) -> f64 {
    let alpha = spec.damping_factor();
    v0 * (-alpha * t).exp() * (1.0 + alpha * t)
}

/// Critically damped current closed form.
pub fn analytical_current_critically_damped(spec: &CapacitorBankSpec, t: f64, v0: f64) -> f64 {
    let alpha = spec.damping_factor();
    (v0 / spec.inductance_h) * t * (-alpha * t).exp()
}

/// Overdamped voltage closed form.
pub fn analytical_voltage_overdamped(spec: &CapacitorBankSpec, t: f64, v0: f64) -> f64 {
    let alpha = spec.damping_factor();
    let omega0 = spec.resonant_frequency();
    let beta = (alpha * alpha - omega0 * omega0).sqrt();
    let s1 = -alpha + beta;
    let s2 = -alpha - beta;
    v0 * (s1 * (s2 * t).exp() - s2 * (s1 * t).exp()) / (s1 - s2)
}

/// Overdamped current closed form.
pub fn analytical_current_overdamped(spec: &CapacitorBankSpec, t: f64, v0: f64) -> f64 {
    let alpha = spec.damping_factor();
    let omega0 = spec.resonant_frequency();
    let beta = (alpha * alpha - omega0 * omega0).sqrt();
    let s1 = -alpha + beta;
    let s2 = -alpha - beta;
    (v0 / (spec.inductance_h * (s1 - s2))) * ((s1 * t).exp() - (s2 * t).exp())
}

/// Return `(v_C(t), i(t))` of the series RLC natural response at time `t`.
pub fn free_response(
    spec: &CapacitorBankSpec,
    t: f64,
    v0: f64,
) -> Result<(f64, f64), FreeResponseError> {
    if t < 0.0 {
        return Err(FreeResponseError::NegativeT);
    }
    Ok(match spec.regime() {
        RlcRegime::Underdamped => (
            analytical_voltage_underdamped(spec, t, v0),
            analytical_current_underdamped(spec, t, v0),
        ),
        RlcRegime::CriticallyDamped => (
            analytical_voltage_critically_damped(spec, t, v0),
            analytical_current_critically_damped(spec, t, v0),
        ),
        RlcRegime::Overdamped => (
            analytical_voltage_overdamped(spec, t, v0),
            analytical_current_overdamped(spec, t, v0),
        ),
    })
}

/// Series RLC capacitor bank with Crank-Nicolson numerical integration.
///
/// Mirrors the Python `scpn_mif_core.lifecycle.capacitor_bank.CapacitorBank`
/// state vector `(v_C, i_L)`. The integrator update solves the 2x2 linear
/// system `(I − Δt/2·A) y_{n+1} = (I + Δt/2·A) y_n + Δt·b` with
/// `A = [[0, -1/C], [1/L, -R/L]]` and `b = [-i_load/C, 0]` via the
/// closed-form 2x2 inverse for bit-true parity with the Python NumPy path.
#[derive(Debug, Clone)]
pub struct CapacitorBank {
    spec: CapacitorBankSpec,
    t: f64,
    v: f64,
    i: f64,
    di_dt: f64,
}

impl CapacitorBank {
    /// Construct a bank, validating that the initial voltage lies in `[0, voltage_max_v]`.
    pub fn new(spec: CapacitorBankSpec, initial_voltage_v: f64) -> Result<Self, ConstructError> {
        require_construct_finite("initial_voltage_v", initial_voltage_v)?;
        if initial_voltage_v > spec.voltage_max_v {
            return Err(ConstructError::ExceedsMax {
                value: initial_voltage_v,
                max: spec.voltage_max_v,
            });
        }
        if initial_voltage_v < 0.0 {
            return Err(ConstructError::NegativeVoltage);
        }
        validate_construct_energy(&spec, initial_voltage_v, 0.0)?;
        Ok(Self {
            spec,
            t: 0.0,
            v: initial_voltage_v,
            i: 0.0,
            di_dt: 0.0,
        })
    }

    /// Read the underlying spec.
    pub fn spec(&self) -> &CapacitorBankSpec {
        &self.spec
    }

    /// Read the current observable state.
    pub fn state(&self) -> CapacitorBankState {
        let capacitor_energy_j = 0.5 * self.spec.capacitance_f * self.v * self.v;
        let inductor_energy_j = 0.5 * self.spec.inductance_h * self.i * self.i;
        CapacitorBankState {
            t: self.t,
            voltage_v: self.v,
            energy_j: capacitor_energy_j + inductor_energy_j,
            capacitor_energy_j,
            inductor_energy_j,
            current_a: self.i,
            di_dt_a_s: self.di_dt,
            discharge_active: self.i.abs() > 1e-9,
            recharge_active: false,
        }
    }

    /// Return the instantaneous short-circuit current bound `|v_C| / sqrt(L / C)`.
    pub fn natural_peak_current_a(&self) -> f64 {
        let characteristic_impedance = (self.spec.inductance_h / self.spec.capacitance_f).sqrt();
        self.v.abs() / characteristic_impedance
    }

    /// Reset the bank to `voltage_v` with zero current and `t = 0`.
    pub fn reset(&mut self, voltage_v: f64) -> Result<(), ConstructError> {
        require_construct_finite("voltage_v", voltage_v)?;
        if voltage_v > self.spec.voltage_max_v {
            return Err(ConstructError::ExceedsMax {
                value: voltage_v,
                max: self.spec.voltage_max_v,
            });
        }
        if voltage_v < 0.0 {
            return Err(ConstructError::NegativeVoltage);
        }
        validate_construct_energy(&self.spec, voltage_v, 0.0)?;
        self.t = 0.0;
        self.v = voltage_v;
        self.i = 0.0;
        self.di_dt = 0.0;
        Ok(())
    }

    /// Advance the bank state by `dt` using Crank-Nicolson.
    ///
    /// `external_load_current_a` is the prescribed current drawn by an
    /// external load attached to the capacitor; zero recovers the natural
    /// response.
    pub fn step(
        &mut self,
        dt: f64,
        external_load_current_a: f64,
    ) -> Result<CapacitorBankState, StepError> {
        require_step_finite("dt", dt)?;
        require_step_finite("external_load_current_a", external_load_current_a)?;
        if dt <= 0.0 {
            return Err(StepError::NonPositiveDt);
        }
        let cap = self.spec.capacitance_f;
        let ind = self.spec.inductance_h;
        let res = self.spec.series_resistance_ohm;
        // A = [[0, -1/C], [1/L, -R/L]]
        let a12 = -1.0 / cap;
        let a21 = 1.0 / ind;
        let a22 = -res / ind;
        let h = dt / 2.0;
        // rhs vector = (I + h·A)·y_n + dt·b where b = [-i_load/C, 0]
        let y_v = self.v;
        let y_i = self.i;
        let rhs_v = y_v + h * a12 * y_i - dt * external_load_current_a / cap;
        let rhs_i = h * a21 * y_v + (1.0 + h * a22) * y_i;
        // lhs = (I − h·A) = [[1, -h·a12], [-h·a21, 1 - h·a22]]
        let l11 = 1.0;
        let l12 = -h * a12;
        let l21 = -h * a21;
        let l22 = 1.0 - h * a22;
        // Closed-form 2x2 inverse: det^{-1}·[[l22, -l12], [-l21, l11]]
        let det = l11 * l22 - l12 * l21;
        let v_next = (l22 * rhs_v - l12 * rhs_i) / det;
        let i_next = (-l21 * rhs_v + l11 * rhs_i) / det;
        let di_dt_next = a21 * v_next + a22 * i_next;
        require_step_finite("voltage_v", v_next)?;
        require_step_finite("current_a", i_next)?;
        require_step_finite("di_dt_a_s", di_dt_next)?;
        validate_step_energy(&self.spec, v_next, i_next)?;
        self.t += dt;
        self.v = v_next;
        self.i = i_next;
        self.di_dt = di_dt_next;
        Ok(self.state())
    }
}

fn require_construct_finite(field: &'static str, value: f64) -> Result<(), ConstructError> {
    if value.is_finite() {
        Ok(())
    } else {
        Err(ConstructError::NonFinite { field })
    }
}

fn validate_construct_energy(
    spec: &CapacitorBankSpec,
    voltage_v: f64,
    current_a: f64,
) -> Result<(), ConstructError> {
    let capacitor_energy_j = 0.5 * spec.capacitance_f * voltage_v * voltage_v;
    require_construct_finite("capacitor_energy_j", capacitor_energy_j)?;
    let inductor_energy_j = 0.5 * spec.inductance_h * current_a * current_a;
    require_construct_finite("inductor_energy_j", inductor_energy_j)?;
    require_construct_finite("energy_j", capacitor_energy_j + inductor_energy_j)
}

fn require_step_finite(field: &'static str, value: f64) -> Result<(), StepError> {
    if value.is_finite() {
        Ok(())
    } else {
        Err(StepError::NonFinite { field })
    }
}

fn validate_step_energy(
    spec: &CapacitorBankSpec,
    voltage_v: f64,
    current_a: f64,
) -> Result<(), StepError> {
    let capacitor_energy_j = 0.5 * spec.capacitance_f * voltage_v * voltage_v;
    require_step_finite("capacitor_energy_j", capacitor_energy_j)?;
    let inductor_energy_j = 0.5 * spec.inductance_h * current_a * current_a;
    require_step_finite("inductor_energy_j", inductor_energy_j)?;
    require_step_finite("energy_j", capacitor_energy_j + inductor_energy_j)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn underdamped_spec() -> CapacitorBankSpec {
        CapacitorBankSpec::new(100e-6, 100e-6, 0.5, 10_000.0, 10.0).unwrap()
    }

    fn overdamped_spec() -> CapacitorBankSpec {
        CapacitorBankSpec::new(100e-6, 100e-6, 10.0, 10_000.0, 10.0).unwrap()
    }

    #[test]
    fn regime_underdamped() {
        assert_eq!(underdamped_spec().regime(), RlcRegime::Underdamped);
    }

    #[test]
    fn regime_overdamped() {
        assert_eq!(overdamped_spec().regime(), RlcRegime::Overdamped);
    }

    #[test]
    fn analytical_at_t0_returns_v0() {
        let spec = underdamped_spec();
        let v0 = 5000.0;
        assert!((analytical_voltage_underdamped(&spec, 0.0, v0) - v0).abs() < 1e-9);
        assert!(analytical_current_underdamped(&spec, 0.0, v0).abs() < 1e-9);
    }

    #[test]
    fn step_rejects_zero_dt() {
        let spec = underdamped_spec();
        let mut bank = CapacitorBank::new(spec, 5000.0).unwrap();
        assert_eq!(bank.step(0.0, 0.0).unwrap_err(), StepError::NonPositiveDt);
    }

    #[test]
    fn step_matches_analytical_underdamped_within_1e_3() {
        let spec = underdamped_spec();
        let v0 = 5000.0;
        let mut bank = CapacitorBank::new(spec, v0).unwrap();
        let dt = 1e-6;
        let n = 100usize;
        for _ in 0..n {
            bank.step(dt, 0.0).unwrap();
        }
        let t = (n as f64) * dt;
        let (v_anal, i_anal) = free_response(&spec, t, v0).unwrap();
        assert!((bank.state().voltage_v - v_anal).abs() / v_anal.abs() < 1e-3);
        assert!((bank.state().current_a - i_anal).abs() / i_anal.abs().max(1e-9) < 1e-3);
    }

    #[test]
    fn free_response_rejects_negative_t() {
        let spec = underdamped_spec();
        assert!(free_response(&spec, -1e-9, 5000.0).is_err());
    }

    #[test]
    fn reset_clears_t_current_didt() {
        let spec = underdamped_spec();
        let mut bank = CapacitorBank::new(spec, 5000.0).unwrap();
        bank.step(1e-5, 0.0).unwrap();
        assert!(bank.state().t > 0.0);
        bank.reset(3000.0).unwrap();
        assert_eq!(bank.state().t, 0.0);
        assert_eq!(bank.state().voltage_v, 3000.0);
        assert_eq!(bank.state().current_a, 0.0);
    }

    #[test]
    fn state_energy_includes_capacitor_and_inductor_storage() {
        let spec = underdamped_spec();
        let mut bank = CapacitorBank::new(spec, 5000.0).unwrap();
        bank.step(1e-6, 0.0).unwrap();
        let state = bank.state();
        let expected_capacitor = 0.5 * spec.capacitance_f * state.voltage_v * state.voltage_v;
        let expected_inductor = 0.5 * spec.inductance_h * state.current_a * state.current_a;
        assert!((state.capacitor_energy_j - expected_capacitor).abs() < 1e-12);
        assert!((state.inductor_energy_j - expected_inductor).abs() < 1e-12);
        assert!((state.energy_j - (expected_capacitor + expected_inductor)).abs() < 1e-12);
        assert!(state.energy_j > state.capacitor_energy_j);
    }

    #[test]
    fn natural_peak_current_matches_characteristic_impedance_bound() {
        let spec = underdamped_spec();
        let bank = CapacitorBank::new(spec, 5000.0).unwrap();
        let expected = 5000.0 / (spec.inductance_h / spec.capacitance_f).sqrt();
        assert!((bank.natural_peak_current_a() - expected).abs() < 1e-12);
    }

    #[test]
    fn construct_rejects_voltage_above_max() {
        let spec = underdamped_spec();
        let err = CapacitorBank::new(spec, spec.voltage_max_v + 1.0).unwrap_err();
        assert!(matches!(err, ConstructError::ExceedsMax { .. }));
    }

    #[test]
    fn spec_rejects_non_finite_parameters_and_derived_energy() {
        assert!(matches!(
            CapacitorBankSpec::new(f64::NAN, 1e-3, 0.1, 1000.0, 10.0),
            Err(crate::types::SpecError::NonFinite {
                field: "capacitance_f"
            })
        ));
        assert!(matches!(
            CapacitorBankSpec::new(1.0e308, 1.0, 0.0, 1.0e154, 0.0),
            Err(crate::types::SpecError::NonFinite {
                field: "max_capacitor_energy_j"
            })
        ));
    }
}

File: julia/SCPNMIFCore/src/capacitor_bank.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN-MIF-CORE — MIF-005 capacitor-bank Julia reference.
#
# OWNED-BY: scpn-mif-core
# CONSUMED-BY: scpn-mif-core
# SYNC-STATE: upstream-pending
# UPSTREAM-PIN: scpn-mif-core@0.0.1
# CONTRACT-TEST: julia/SCPNMIFCore/test/runtests.jl
# TRACKED-ISSUE: docs/internal/upstream_contracts/03_scpn_control.md#con-c2-capacitorbank-state-model
# LAST-SYNCED: 2026-06-04T0000

"""Series RLC natural-response classification."""
@enum RLCRegime OVERDAMPED CRITICALLY_DAMPED UNDERDAMPED

"""Immutable physical and operational specification of a capacitor bank."""
struct CapacitorBankSpec
    capacitance_F::Float64
    inductance_H::Float64
    series_resistance_ohm::Float64
    voltage_max_V::Float64
    recharge_power_kW::Float64

    function CapacitorBankSpec(
        capacitance_F::Real,
        inductance_H::Real,
        series_resistance_ohm::Real,
        voltage_max_V::Real,
        recharge_power_kW::Real,
    )
        cap = _validate_capacitor_positive("capacitance_F", capacitance_F)
        ind = _validate_capacitor_positive("inductance_H", inductance_H)
        res = _validate_capacitor_non_negative("series_resistance_ohm", series_resistance_ohm)
        vmax = _validate_capacitor_positive("voltage_max_V", voltage_max_V)
        recharge = _validate_capacitor_non_negative("recharge_power_kW", recharge_power_kW)
        return new(cap, ind, res, vmax, recharge)
    end
end

"""Observable capacitor-bank state."""
struct CapacitorBankState
    t::Float64
    voltage_V::Float64
    energy_J::Float64
    capacitor_energy_J::Float64
    inductor_energy_J::Float64
    current_A::Float64
    di_dt_A_s::Float64
    discharge_active::Bool
    recharge_active::Bool
end

"""Stateful Crank-Nicolson integrator for the MIF-005 series RLC bank."""
mutable struct CapacitorBank
    spec::CapacitorBankSpec
    t::Float64
    voltage_V::Float64
    current_A::Float64
    di_dt_A_s::Float64

    function CapacitorBank(spec::CapacitorBankSpec, initial_voltage_V::Real = 0.0)
        voltage = _validate_capacitor_non_negative("initial voltage", initial_voltage_V)
        voltage <= spec.voltage_max_V ||
            throw(ArgumentError("initial voltage $(voltage) V exceeds bank max $(spec.voltage_max_V) V"))
        return new(spec, 0.0, voltage, 0.0, 0.0)
    end
end

"""Return damping factor `alpha = R / (2L)`."""
damping_factor(spec::CapacitorBankSpec)::Float64 =
    spec.series_resistance_ohm / (2.0 * spec.inductance_H)

"""Return undamped resonant frequency `omega_0 = 1 / sqrt(LC)`."""
resonant_frequency(spec::CapacitorBankSpec)::Float64 =
    1.0 / sqrt(spec.inductance_H * spec.capacitance_F)

"""Return critical resistance `R_crit = 2 * sqrt(L / C)`."""
critical_resistance(spec::CapacitorBankSpec)::Float64 =
    2.0 * sqrt(spec.inductance_H / spec.capacitance_F)

"""Return the series RLC damping regime."""
function regime(spec::CapacitorBankSpec)::RLCRegime
    resistance = spec.series_resistance_ohm
    critical = critical_resistance(spec)
    if isapprox(resistance, critical; rtol = 1e-9, atol = 1e-15)
        return CRITICALLY_DAMPED
    end
    return resistance < critical ? UNDERDAMPED : OVERDAMPED
end

"""Underdamped voltage closed form."""
function analytical_voltage_underdamped(spec::CapacitorBankSpec, t::Real, v0::Real)::Float64
    time = _validate_capacitor_non_negative("t", t)
    initial = _require_finite_capacitor("v0", v0)
    alpha = damping_factor(spec)
    omega0 = resonant_frequency(spec)
    omega_d = sqrt(omega0^2 - alpha^2)
    return initial * exp(-alpha * time) *
        (cos(omega_d * time) + (alpha / omega_d) * sin(omega_d * time))
end

"""Underdamped current closed form."""
function analytical_current_underdamped(spec::CapacitorBankSpec, t::Real, v0::Real)::Float64
    time = _validate_capacitor_non_negative("t", t)
    initial = _require_finite_capacitor("v0", v0)
    alpha = damping_factor(spec)
    omega0 = resonant_frequency(spec)
    omega_d = sqrt(omega0^2 - alpha^2)
    return (initial / (omega_d * spec.inductance_H)) * exp(-alpha * time) * sin(omega_d * time)
end

"""Critically damped voltage closed form."""
function analytical_voltage_critically_damped(spec::CapacitorBankSpec, t::Real, v0::Real)::Float64
    time = _validate_capacitor_non_negative("t", t)
    initial = _require_finite_capacitor("v0", v0)
    alpha = damping_factor(spec)
    return initial * exp(-alpha * time) * (1.0 + alpha * time)
end

"""Critically damped current closed form."""
function analytical_current_critically_damped(spec::CapacitorBankSpec, t::Real, v0::Real)::Float64
    time = _validate_capacitor_non_negative("t", t)
    initial = _require_finite_capacitor("v0", v0)
    alpha = damping_factor(spec)
    return (initial / spec.inductance_H) * time * exp(-alpha * time)
end

"""Overdamped voltage closed form."""
function analytical_voltage_overdamped(spec::CapacitorBankSpec, t::Real, v0::Real)::Float64
    time = _validate_capacitor_non_negative("t", t)
    initial = _require_finite_capacitor("v0", v0)
    alpha = damping_factor(spec)
    omega0 = resonant_frequency(spec)
    beta = sqrt(alpha^2 - omega0^2)
    s1 = -alpha + beta
    s2 = -alpha - beta
    return initial * (s1 * exp(s2 * time) - s2 * exp(s1 * time)) / (s1 - s2)
end

"""Overdamped current closed form."""
function analytical_current_overdamped(spec::CapacitorBankSpec, t::Real, v0::Real)::Float64
    time = _validate_capacitor_non_negative("t", t)
    initial = _require_finite_capacitor("v0", v0)
    alpha = damping_factor(spec)
    omega0 = resonant_frequency(spec)
    beta = sqrt(alpha^2 - omega0^2)
    s1 = -alpha + beta
    s2 = -alpha - beta
    return (initial / (spec.inductance_H * (s1 - s2))) * (exp(s1 * time) - exp(s2 * time))
end

"""Return `(v_C(t), i_L(t))` for the natural series RLC response."""
function free_response(spec::CapacitorBankSpec, t::Real, v0::Real)::Tuple{Float64, Float64}
    time = _validate_capacitor_non_negative("t", t)
    initial = _require_finite_capacitor("v0", v0)
    current_regime = regime(spec)
    if current_regime == UNDERDAMPED
        return (
            analytical_voltage_underdamped(spec, time, initial),
            analytical_current_underdamped(spec, time, initial),
        )
    elseif current_regime == CRITICALLY_DAMPED
        return (
            analytical_voltage_critically_damped(spec, time, initial),
            analytical_current_critically_damped(spec, time, initial),
        )
    end
    return (
        analytical_voltage_overdamped(spec, time, initial),
        analytical_current_overdamped(spec, time, initial),
    )
end

"""Return an immutable observable snapshot of the bank."""
function state(bank::CapacitorBank)::CapacitorBankState
    capacitor_energy = 0.5 * bank.spec.capacitance_F * bank.voltage_V^2
    inductor_energy = 0.5 * bank.spec.inductance_H * bank.current_A^2
    return CapacitorBankState(
        bank.t,
        bank.voltage_V,
        capacitor_energy + inductor_energy,
        capacitor_energy,
        inductor_energy,
        bank.current_A,
        bank.di_dt_A_s,
        abs(bank.current_A) > 1e-9,
        false,
    )
end

"""Reset a bank to `voltage_V` with zero current and time."""
function reset!(bank::CapacitorBank, voltage_V::Real = 0.0)::CapacitorBankState
    voltage = _validate_capacitor_non_negative("reset voltage", voltage_V)
    voltage <= bank.spec.voltage_max_V ||
        throw(ArgumentError("reset voltage $(voltage) V exceeds bank max $(bank.spec.voltage_max_V) V"))
    bank.t = 0.0
    bank.voltage_V = voltage
    bank.current_A = 0.0
    bank.di_dt_A_s = 0.0
    return state(bank)
end

"""Advance the bank by `dt` seconds using Crank-Nicolson."""
function step!(
    bank::CapacitorBank,
    dt::Real,
    external_load_current_A::Real = 0.0,
)::CapacitorBankState
    step_s = _validate_capacitor_positive("dt", dt)
    load_current = _require_finite_capacitor("external_load_current_A", external_load_current_A)
    cap = bank.spec.capacitance_F
    ind = bank.spec.inductance_H
    res = bank.spec.series_resistance_ohm

    a12 = -1.0 / cap
    a21 = 1.0 / ind
    a22 = -res / ind
    h = step_s / 2.0

    rhs_v = bank.voltage_V + h * a12 * bank.current_A - step_s * load_current / cap
    rhs_i = h * a21 * bank.voltage_V + (1.0 + h * a22) * bank.current_A

    l11 = 1.0
    l12 = -h * a12
    l21 = -h * a21
    l22 = 1.0 - h * a22
    det = l11 * l22 - l12 * l21

    v_next = (l22 * rhs_v - l12 * rhs_i) / det
    i_next = (-l21 * rhs_v + l11 * rhs_i) / det
    di_dt_next = a21 * v_next + a22 * i_next

    bank.t += step_s
    bank.voltage_V = v_next
    bank.current_A = i_next
    bank.di_dt_A_s = di_dt_next
    return state(bank)
end

function Base.getproperty(bank::CapacitorBank, name::Symbol)
    if name == :state
        return state(bank)
    end
    return getfield(bank, name)
end

function _require_finite_capacitor(name::String, value::Real)::Float64
    numeric = Float64(value)
    isfinite(numeric) || throw(ArgumentError("$name must be finite"))
    return numeric
end

function _validate_capacitor_positive(name::String, value::Real)::Float64
    numeric = _require_finite_capacitor(name, value)
    numeric > 0.0 || throw(ArgumentError("$name must be strictly positive"))
    return numeric
end

function _validate_capacitor_non_negative(name::String, value::Real)::Float64
    numeric = _require_finite_capacitor(name, value)
    numeric >= 0.0 || throw(ArgumentError("$name must be non-negative"))
    return numeric
end
